keel-core/src/main/kotlin/com/netflix/spinnaker/keel/actuation/EnvironmentConstraintRunner.kt

```diff
@@ -0,0 +1,250 @@
+            checkStatefulConstraints(envContext.artifact, envContext.deliveryConfig, version, envContext.environment)
+
+        if (passesConstraints) {
+          queueForApproval(envContext.deliveryConfig, envContext.artifact, version, envContext.environment.name)
+        }
+      }
+  }
+
+  /**
+   * A container to hold all the context we need for evaluating environment constraints
+   */
+  data class EnvironmentContext(
+    val deliveryConfig: DeliveryConfig,
+    val environment: Environment,
+    val artifact: DeliveryArtifact,
+    val versions: List<String>,
+    val vetoedArtifacts: Map<String, EnvironmentArtifactVetoes>
+  ) {
+    fun hasNoConstraints(): Boolean =
+      environment.constraints.isEmpty()
+
+    fun vetoedVersions(): Set<String> =
+      (vetoedArtifacts[envPinKey()]?.versions)
+        ?: emptySet()
+
+    private fun envPinKey(): String =
+      "${environment.name}:${artifact.name}:${artifact.type.name.toLowerCase()}"
+  }
+
+  /**
+   * Queues a version for approval if it's not already approved in the environment.
+   * [EnvironmentPromotionChecker] handles actually approving a version for an environment.
+   */
+  private fun queueForApproval(
+    deliveryConfig: DeliveryConfig,
+    artifact: DeliveryArtifact,
+    version: String,
+    targetEnvironment: String
+  ) {
+    val latestVersion = repository
+      .latestVersionApprovedIn(deliveryConfig, artifact, targetEnvironment)
+    if (latestVersion != version) {
+      log.debug("Queueing version $version of ${artifact.type} artifact ${artifact.name} in environment $targetEnvironment for approval")
+      repository.queueAllConstraintsApproved(deliveryConfig.name, targetEnvironment, version)
+    }
+  }
+
+  fun checkStatelessConstraints(
+    artifact: DeliveryArtifact,
+    deliveryConfig: DeliveryConfig,
+    version: String,
+    environment: Environment
+  ): Boolean =
+    checkImplicitConstraints(implicitStatelessEvaluators, artifact, deliveryConfig, version, environment) &&
+      checkConstraints(statelessEvaluators, artifact, deliveryConfig, version, environment)
+
+  fun checkStatefulConstraints(
+    artifact: DeliveryArtifact,
+    deliveryConfig: DeliveryConfig,
+    version: String,
+    environment: Environment
+  ): Boolean =
+    checkImplicitConstraints(implicitStatefulEvaluators, artifact, deliveryConfig, version, environment) &&
+      checkConstraints(statefulEvaluators, artifact, deliveryConfig, version, environment)
+
+  /**
+   * Checks constraints for a list of evaluators.
+   * Evaluates the constraint for every environment passed in.
+   * @return true if all constraints pass
+   */
+  private fun checkImplicitConstraints(
+    evaluators: List<ConstraintEvaluator<*>>,
+    artifact: DeliveryArtifact,
+    deliveryConfig: DeliveryConfig,
+    version: String,
+    environment: Environment
+  ): Boolean {
+    return if (evaluators.isEmpty()) {
```

Comment:
There's no need to special case the empty list: `evaluators.all{ ... }` returns `true` when `evaluators` is empty.

Reply:
nice! will do